File: Recognized Patterns/Merge Sort/QJs/merge-sort.js
/**
 * Break up the array into haves until you have arrays that are empty or
 * have one element
 * 
 * Once tou have smaller sorted arrays, merge those arrays with other sorted
 * arrays until you are back at the full length og the array
 * 
 * Once the array has been merged back together, return the merged array.
*/


function merge(arr1, arr2) {
    let n = arr1.length, m = arr2.length
    let i=0, j=0
    let result = []
    while(i<n && j<m) {
        if(arr1[i]<=arr2[j]) {
            result[result.length]=arr1[i]
            i++
        } else if(arr2[j] <= arr1[i]) {
            result[result.length]=arr2[j]
            j++
        }
    }
    while(i<n) {
        result[result.length]=arr1[i]
        i++
    }
    while(j<m) {
        result[result.length]=arr2[j]
        j++
    }

    return result
}

function mergeSort(arr) {
    // base case
    if(arr.length <= 1) return arr

    let mid = Math.floor(arr.length/2)
    let left = mergeSort(arr.slice(0, mid))
    let right = mergeSort(arr.slice(mid))

    return merge(left, right)
}

log = (val) => console.log(val)

log(merge([1,2,3],[3,6,7]))
log(mergeSort([1,4,3,5,32,6,7,6,8,7,66,5,4,3,4,6]))

File: Practice Library/Any Language/Sorting and Searching/Merge Sort/QJs/merge_sort.py
import math

def merge_sort(arr: list):
    if(len(arr)<=1): return arr
    mid = math.floor(len(arr)//2)
    a=[]
    b=[]
    for i, v in enumerate(arr):
        a.append(v)
        if(i==mid-1): break
    j=mid
    while(j<len(arr)):
        b.append(arr[j])
        j+=1
    left = merge_sort(a)
    right = merge_sort(b)
    return merge(left,right)


def merge(a: list, b: list):
    i=0
    j=0
    result = []
    while(i<=len(a)) and (j<len(b)):
        if(a[i]<b[j]):
            result.append(a[i])
            i+=1
        else:
            result.append(b[j])
            j+=1
    
    while(i<len(a)):
        result.append(a[i])
        i+=1
    
    while(j<len(b)):
        result.append(b[j])
        j+=1
    
    return result



print(merge_sort([9,8,7,6,5,4]))